package/MDAnalysis/coordinates/base.py

```diff
@@ -1663,7 +1663,11 @@ class SingleFrameReaderBase(ProtoReader):
     def __init__(self, filename, convert_units=True, n_atoms=None, **kwargs):
         super(SingleFrameReaderBase, self).__init__()
 
-        self.filename = filename
+        if isinstance(filename, NamedStream):
+            self.filename = filename
+        else:
+            self.filename = str(filename)
```

Comment:
Here errors if we use `self.filename = str(filename)` in the `test_mmtf.py`
The lead to errors similar to this one:

`FileNotFoundError: [Errno 2] No such file or directory: '<mmtf.api.mmtf_reader.MMTFDecoder object at 0x135416780>'`

I managed to fix the error by adding:
`else:`
`    self.filename = str(filename)`
`    if "MMTF" in self.filename:`
`        self.filename = filename`

basically putting a specific condition for MMTF cases, not the cleanest approach, but maybe this helps @hmacdope fixing the error in a better way :)

Reply:
I tried to look more into it now, the `"MMTF"` is a specific case, the other errors appear mainly because of the parmed parser which has both problems with the `base.py` in the `coordinates` and `topology`. It is possible to use in the `base.py` in the `coordinates`:

`elif hasattr(filename, 'topology'):`
`    self.filename = filename`

This helps to fix the errors, although it is quite similar as to just directly using `self.filename = filename` without any `if/elif/else` conditions. For the `base.py` in `topology` I will try to figure out if there is any option to fix it, since again having there only `self.filename = filename` also fixes the errors with the `parmed` parser

Reply:
@talagayev I think the `FileNotFoundError` for mmtf above is because instead of a filename, sometimes a MMTFDecoder object is encountered.  I think what you've got is fixing it, but it might be clearer to have a block like:

```python
if isinstance(filename, NamedStream):
    self.filename = filename
elif 'mmtf' in str(filename.__class__):
    # mmtf case, where we've avoided a mmtf import in detection
    self.filename = filename
```

Where you've enumerated all the special cases explicitly in each `elif` branch, rather than hiding some special cases inside other branches

Reply:
Ah I see, yes the option that I got worked, but was definitely not the best option, but yes was connected with mmtf and yes  `elif 'mmtf' in str(filename.__class__):` looks much better :)